Implement optional per player mobspawning

kickash32 opened the pull request:
Note this Patch essentially reimplements #2127 if disabled.
Implements requests from #245 and #1653.

If enabled, the chunks(with range taken from spigot.yml's mob-spawn-range) around players who have the sufficient amount of mobs of a type (limit taken from bukkit.yml's mob limits), will not have chunks around them eligible to spawn mobs of the same type. Additionally the global limit is ignored (although in most circumstances the number of mobs will be the same if mob-spawn-range is 8).

Benefits include more balanced/fair mob spawning, a player on the other side of the world will not be able to affect mob spawning to other players, etc. Essentially makes mob spawning behave as each player was in single player and so it includes all the additional benefits that come with that.

~~It works by creating a map of player, entity type to counts which is recreated every tick and updated as mobs are spawned. For each potential chunk available for spawning, the map is queried with all the players in range to determine how many mobs should be spawned in the chunk.~~ See below.

kickash32 committed:
Create 0402-implement-optional-per-player-mob-spawns.patch

kickash32 committed:
ignore global limit, reduce work when disabled

electronicboy committed:
Updated Upstream (CraftBukkit)

Upstream has released updates that appears to apply and compile correctly.
This update has not been tested by PaperMC and as with ANY update, please do your own testing

CraftBukkit Changes:
9e95da12 SPIGOT-5052: PlayerPortalEvent not called when nether is disabled
28ea3fd1 SPIGOT-5062: end_portal on an non-default end world crashes the server
4a07dc69 SPIGOT-5058: ClassCastException when getting InventoryHolder of lectern when it's destroyed

kickash32 committed:
Stop natural spawning after limit has been reached (#2127)

Spottedleaf committed:
Update upstream & fix some chunk related issues (#2177)

* Updated Upstream (Bukkit/CraftBukkit)

Upstream has released updates that appears to apply and compile correctly.
This update has not been tested by PaperMC and as with ANY update, please do your own testing

Bukkit Changes:
45690fe9 SPIGOT-5047: Correct slot types for 1.14 inventories

CraftBukkit Changes:
4090d01f SPIGOT-5047: Correct slot types for 1.14 inventories
e8c08362 SPIGOT-5046: World#getLoadedChunks returning inaccessible cached chunks.
d445af3b SPIGOT-5067: Add item meta for 1.14 spawn eggs

* Bring Chunk load checks in-line with spigot

As of the last upstream merge spigot now checks ticket level status
when returning loaded chunks for a world from api. Now our checks
will respect that decision.

* Fix spawn ticket levels

Vanilla would keep the inner chunks of spawn available for ticking,
however my changes made all chunks non-ticking. Resolve by changing
ticket levels for spawn chunks inside the border to respect this
behavior.


* Make World#getChunkIfLoadedImmediately return only entity ticking chunks

Mojang appears to be using chunks with level > 33 (non-ticking chunks)
as cached chunks and not actually loaded chunks.

* Bring all loaded checks in line with spigot

Loaded chunks must be at least border  chunks, or level <= 33

Spottedleaf committed:
Add chunkinfo command

Used so that server owners can tell what types of chunks are loaded
without requiring to use a tool to analyse a debug report

Spottedleaf committed:
Change format

wizjany committed:
Use the entity registry id instead of translation key when posting PreCreatureSpawnEvent and PreSpawnerSpawnEvent

Spottedleaf committed:
Show blockstate location if we failed to read it

Spottedleaf committed:
Proper paper commenting

Spottedleaf committed:
fix exception message

kickash32 committed:
Remove patch since it was reimplemented

kickash32 committed:
fix null pointer exception

kickash32 committed:
Merge pull request #3 from PaperMC/ver/1.14

Ver/1.14

kickash32 committed:
Revert "Remove patch since it was reimplemented"

This reverts commit c28619ae981c377a96cdae1fc99614b3993179a1.

kickash32 committed:
Delete 0398-Actually-Limit-Natural-Spawns-To-Limit.patch

kickash32 committed:
fix race condition

Spottedleaf commented on Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,244 @@
++    public Stream<EntityPlayer> getPlayers(ChunkCoordIntPair chunkcoordintpair, int range) {
++        int blockRange = ((range << 4) * (range << 4));
++        return this.playerMap.a(chunkcoordintpair.pair()).filter((entityplayer) ->
++            (!entityplayer.isSpectator() && a(chunkcoordintpair, (Entity) entityplayer) < blockRange));
```
obfhelpers

Black-Hole replied:
@Spottedleaf wasn't talking about comment lines here.
You need an OPFHELPER if you call obfuscated methods like a(). So here for PlayerMap.a() and PlayerChunkMap.a()

kickash32 replied:
Made some changes, please take a look.

Black-Hole replied:
It looks fine, but OBFHELPERS won't trigger a conflict if there is an empty line between the OBFHELPER and the obfuscated method.

Black-Hole commented on Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,244 @@
+ import java.util.function.BooleanSupplier;
+ import java.util.function.Function;
+ import java.util.function.Supplier;
++import java.util.stream.Stream; // Paper
```
This is used only once in the file, so you could use it fully qualified below.

kickash32 replied:
Done.

Black-Hole commented on Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,244 @@
+                 if (enumcreaturetype != EnumCreatureType.MISC && this.getChunkProvider().a(entity)) {
++                    filteredEntities.add(entity); // Paper
+                     object2intmap.computeInt(enumcreaturetype, (enumcreaturetype1, integer) -> {
+                         return 1 + (integer == null ? 0 : integer);
```
This is a Object2IntMap, so integer won't be null, but using the default value 0.

kickash32 replied:
Not sure what you are referring to, but lines 196-197 are mojang code

kickash32 committed:
Use player UUID and minor refactoring

kickash32 committed:
add OBFHELPERs

kickash32 committed:
fix mapping issue

kickash32 committed:
Merge pull request #4 from PaperMC/ver/1.14

merge upstream 1.14.3

kickash32 committed:
1.14.3 compatability and refactoring

kashike commented on Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,241 @@
+     private boolean ticking;
+     @Nullable
+     private final MobSpawnerTrader mobSpawnerTrader;
++    private Map<UUID, Map<EnumCreatureType, Integer>> playerMobTypeMap; // Paper
```
`Map<UUID, Object2IntMap<EnumCreatureType>>`?

Spottedleaf replied:
enummap better

kickash32 replied:
I used EnumMap

kickash32 committed:
Merge pull request #7 from PaperMC/ver/1.14

update to Ver/1.14.4

kickash32 committed:
reorder

kickash32 committed:
Update for 1.14.4

kickash32 committed:
remove stream

kickash32 committed:
remove more stream

kickash32 committed:
add OBFHELPER and fix comments

Spottedleaf commented on Spigot-Server-Patches/0406-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,259 @@
++
++            int i = distanceBetween(chunkcoordintpair, entityplayer);
++
++            if(i > this.viewDistance) { iterator.remove(); }
```
iterator remove on the uncopied playermap

kickash32 replied:
I updated it to return a copy of the keyset so this modification should be safe

Spottedleaf commented on Spigot-Server-Patches/0406-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,259 @@
+ 
++    // Paper start
++    public void updatePlayerMobTypeMap(List<Entity> entities) {
++        if (this.playerMobTypeMap == null) { return; }
```
the map needs to be cleared at this point, you're accumulating data

paul-maxime replied:
Unless I am mistaken, `playerMobTypeMap` is cleared every tick, or at least each time the spawnmob method is called.

Spottedleaf replied:
yeah sorry it is

kickash32 committed:
Return copy of keyset instead of keyset

kickash32 committed:
ignore spectators and players who don't affect spawning

zachbr reviewed (approved):
I have some minor quibbles about some formatting but overall it looks good.

kickash32 committed:
fix comments, move diffence calc outside of loop

Also some minor refactoring of new methods

kickash32 committed:
reuse spawnedMob list and minimize diff on SpawnerCreature.java

kickash32 committed:
store mobcounts in players and use LeakyResettableList

kickash32 committed:
Use consumer instead of list and fix count reset bug

kickash32 committed:
remove unneeded imports

kickash32 committed:
reduce diff and optimize updatePlayerMobCounts

kickash32 committed:
Merge remote-tracking branch 'upstream/ver/1.14' into mob-spawning-rework

kickash32 committed:
reorder

kickash32 committed:
restore original functionality of World#l()

Instead use `l(true)` for future compatability where `l()` may be called in other places.

kickash32 committed:
update comments and rename method l -> countMobs

Spottedleaf committed:
Optimize

Adds a distance map where players are tracked to coordinates in which
they are close to. The class contains a validatePlayer() method
which was used for testing.

The class uses pooled hash tables to reduce object churning, and was
tested via the following:

https://gist.github.com/Spottedleaf/a93bb7a8993d6ce142d3efc5932bf573

With the distance map, looking up players in a specific location will be
approximately constant time, which makes this patch scale much better with
more players, at the cost of some performance at lower playercounts.

Spottedleaf committed:
Do not immediately discard mob counts for mobs already in the world

Spottedleaf committed:
Remove duplicate methods

kickash32 committed:
Merge pull request #8 from Spottedleaf/mob-spawning-rework

Optimize

Spottedleaf committed:
Fix teleportation

kickash32 committed:
Merge pull request #9 from Spottedleaf/mob-spawning-rework

Fix teleportation

Spottedleaf commented on Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,806 @@
++                }
++                worldMobCount = this.world.countMobs(true);
++                // update distance map
++                this.playerChunkMap.playerMobDistanceMap.update(this.world.players, this.playerChunkMap.viewDistance);
```
this should be moved before the mob count reset

also move the stop timing handler after the update method

kickash32 committed:
reorder some code

kickash32 committed:
Merge remote-tracking branch 'upstream/ver/1.14' into mob-spawning-rework

kickash32 committed:
resolve merge conflicts

Spottedleaf commented on Spigot-Server-Patches/0412-implement-optional-per-player-mob-spawns.patch:
```diff
@@ -0,0 +1,806 @@
+         this.m = supplier;
+         this.n = new VillagePlace(new File(this.x, "poi"), datafixer);
+         this.setViewDistance(i);
++        this.playerMobDistanceMap = this.world.paperConfig.perPlayerMobSpawns ? new com.destroystokyo.paper.util.PlayerMobDistanceMap() : null; // Paper
```
missing //paper enclosure

kickash32 replied:
I don't think its required for the timing handler file...

kickash32 committed:
Merge remote-tracking branch 'upstream/ver/1.14' into mob-spawning-rework

kickash32 committed:
fix merge

electronicboy committed:
Rebuild patches
